### ISSUE TITLE
Uninstall of python-dotenv does not (really) work

### ISSUE THREAD

**Issue by soraphis** [pypa/pipenv#5652]
### Issue description

I also tried reaching out on [stackoverflow](https://stackoverflow.com/questions/75999036/pipenv-python-dotenv-reappearing-after-every-pipenv-install?noredirect=1#comment134040336_75999036), but at this point. it feels like a bug

version: pipenv, version 2023.3.20

I try to uninstall python-dotenv. Should be easy, right? It's not in my Pipfile, and in my pipenv-graph it's not listed as a transitive dependency.

Following this series of commands:

<details>

  ```
$ echo "graph"; pipenv graph | grep python-dotenv; echo "Pipfile.lock"; cat Pipfile.lock | grep python-dotenv; echo "Pipfile"; cat Pipfile | grep python-dotenv   
  graph
  Loading .env environment variables...
  python-dotenv==1.0.0
  Pipfile.lock
          "python-dotenv": {
  Pipfile
  $ pipenv uninstall python-dotenv
  Loading .env environment variables...
  Removing python-dotenv from Pipfile.lock...
  Uninstalling python-dotenv...
  Found existing installation: python-dotenv 1.0.0
  Uninstalling python-dotenv-1.0.0:
    Successfully uninstalled python-dotenv-1.0.0
  
  Locking [packages] dependencies...
  Building requirements...
  Resolving dependencies...
  ✔ Success!
  Locking [dev-packages] dependencies...                                                                                                                                                                                                                               
  Building requirements...
  Resolving dependencies...
  ✔ Success!
  Updated Pipfile.lock (9362bc95e99217f2f5cac841a8b99b7b491e5c52886bf094ff10df3874969e05)!                                                                                                                                                                             
  $ echo "graph"; pipenv graph | grep python-dotenv; echo "Pipfile.lock"; cat Pipfile.lock | grep python-dotenv; echo "Pipfile"; cat Pipfile | grep python-dotenv   
  graph
  Loading .env environment variables...
  Pipfile.lock
          "python-dotenv": {
  Pipfile

```

</details>

notice, how it's NOT deleted from the Pipfile.lock?, well, if i now run "pipenv install", I went full circle and have it back in my pipenv graph.

I thought removing it manually from the pipfile lock would work, but after running "pipenv lock" (or any other command that recreates the lockfile) it's back in there again, even if it was not present in the Pipfile nor pipenv graph.

### Expected result

the package being actually removed, and not reappearing again

-------------------------------------------------------------------------------

pipenv --support is way to large to paste here
[pipenv__support.txt](https://github.com/pypa/pipenv/files/11225798/pipenv__support.txt)

**Comment by matteius** [comment]
@soraphis Ok I spent a bit debugging this, and it seems it is because you have `django-dotenv = "*"` in your `Pipfile` which certainly requires `python-dotenv`  -- I am not sure why its showing up as a top level item in the graph though.

**Comment by matteius** [comment]
Sorry, I was slightly off base in my last response, the real reason its included is `uvicorn`:  https://github.com/encode/uvicorn/blob/master/pyproject.toml#L41

**Comment by soraphis** [comment]
Thanks for taking the time and looking into it. What I don't really get is: why isn't showing up in the graph? Other dependencies are also listed multiple times if required by multiple packages...

**Comment by derula** [comment]
For the record, I tested this the other day out of curiosity. PDM also installs `python-dotenv` when importing from this `Pipfile`, but doesn't list it in `pdm list --graph` at all, which I thought was very strange.

**Comment by matteius** [comment]
The new version of pipdeptree appears to fix this, I have a PR out to update pipenv.

**Comment by matteius** [comment]
Well it kind of fixes it -- it shows up in the graph, just not under uvicorn, which seems odd.

**Comment by derula** [comment]
Whether it appears in the graph is kind of secondary, I think. The real problem here is that:
- `python_dotenv` is an _optional_ dependency of `uvicorn`
- `python_dotenv` and `dotenv` are incompatible

Thus:
- Nothing strictly requires `python_dotenv` (at least `uvicorn` doesn't, I haven't checked all other packages)
- `python_dotenv` is installed anyway (for what reason?)
- `dotenv` package breaks / may break as a result

IIRC, `pipenv graph` always showed `python_dotenv` as a top-level dependency, not under `uvicorn`.

Pure speculation: maybe the resolver includes all optional dependencies, thus `python_dotenv` is installed, but the dependency tree does _not_ consider optional dependencies, thus it is not listed as a dependency of any other package.

**Comment by matteius** [comment]
Actually that part is easy to explain @derula --

`uvicorn = {extras = ["standard"], version = "*"}`

`extras = ["standard"]` does tell it to install all of: https://github.com/encode/uvicorn/blob/master/pyproject.toml#L38-L46

**Comment by derula** [comment]
Ah, okay, I must have missed that. Then I guess there are 2 separate problems:
- @soraphis shouldn't be using the `standard` extra because that will cause `python-dotenv` to be installed, conflicting with `dotenv`
- `python-dotenv` (maybe all optional deps?) not listed as sub-dependency of `uvicorn`

**Comment by matteius** [comment]
> python-dotenv (maybe all optional deps?) not listed as sub-dependency of uvicorn

@derula that should be reported upstream to `pipdeptree`

**Comment by derula** [comment]
> @derula that should be reported upstream to `pipdeptree`

Upstream ticket has existed upstream for over 4 years: tox-dev/pipdeptree/issues/107
PR has since been created, diverged, and closed: tox-dev/pipdeptree/pull/138
Seems like it won't be fixing itself :)
